react-intl i18n internationalization and pluralization integration

SilentCicero opened the pull request:
This integrates Yahoo's `react-intl` internationalization and pluralization package into the `react-boilerplate`.

This PR is a first pass at internationalization and integration. We can further this with multi-language support, in-app redux support (so language can be changed on the fly) and other features we deep not too bulky.

This PR adds the `IntlProvider` object to the main `app.js` in both the `app` and `templates` folder.

It also adds the `FormattedMessage` and `FormattedNumber` components to all text and numbers used within the example `app` and the `internals/templates` containers/components.

`react-intl` titles each formatted message and number with an ID. The ID naming scheme I've choosen is similar to that of the `constants` naming scheme used in the example app and scaffolding with some slight modifications for component thinking and dot notation. The ID system is used to build language JSON files that can be front-loaded in the `IntlProvider`. Having all language in JSON files allows others to mak language support easier.

Example ID use:
```js
<FormattedMessage
  id="boilerplate.components.Footer.license.message"
  defaultMessage={`
    This project is licensed under the MIT license.
  `}
/>
```

Some intl ID examples:
```
boilerplate.containers.HomePage.start_project.header
boilerplate.containers.HomePage.start_project.message
boilerplate.containers.FeaturePage.features.header
boilerplate.components.Footer.license.message
boilerplate.components.Footer.author.message
```

- Here you can see the full app name being used. The container or component it is picking out. The capitalized exact component name. The text and context that is being referred too.
- This way, language files can point exactly to the component or container.
- Component pointers are exact names. While other sub-text and sections are lowercase and using underscores.

Pitfalls:
(1) `react-intl` has a few snags that should be made aware of in the docs.

One is that for enzyme testing, template literal spacing, if even slightly incorrect, between what FormattedMessage is being tested, will not equal the same text spacing in enzyme.

To prevent this, exact spacing should be used, or unused spacing should be removed. Otherwise the components wont match in enzyme thus causing an error.

(2) Another pitfall is that scrollSpy among others require the `IntlProvider` object to be wrapped around the `mount` enzyme tests. I'm less worried about this. But it is a requirement.

(3) React-native support, see: https://github.com/yahoo/react-intl/issues/119 -- not sure if this is our concern right now.

I think these pitfalls are minor, as I believe the benefits to internationalization outweigh the negatives here.

We may want to rename a few components here or ID's. And consider the pitfalls that I have found so far.

Thoughts?

mxstbr commented:
@SilentCicero thanks so much for doing this—this is a very solid PR, it looks great! :+1: :fire: :+1: I'm not too worries about the pitfalls, they sound absolutely fine.

Question: Right now, in `app.js`, we hardcode the locale `<IntlProvider locale="en">`. How would changing it based on the language of the browser/OS work, and a language switcher?

One small nitpick: Could you revert the package.json whitespace changes? I like it better the concise way, I think it's more easily readable. Thanks!

SilentCicero commented:
@mxstbr `package.json` should be fixed now with original white spaces.

Re: `app.js` locale settings. Considering that I'm not sure how we want to proceed on the multi-language support (as I haven't fully vetted the `react-intl-redux` package), I think we should leave it hard coded in English for now (as that is what we designed the boilerplate in initially). 
    The idea is that, the boilerplate can now properly support multiple languages. The infrastructure piece is there for that. I think app devs can decide how they want to support multiple languages in the meantime, until we decide on how we want to support multiple languages.

I think @benjaminlgt-littlelives can take it from here to generate the language files and plug those into the boilerplate itself.

At least now this boilerplate is build with i18n support in mind. In my opinion that was the last critical piece to make this thing a killer boiler.

Thoughts?

mxstbr commented:
> The idea is that, the boilerplate can now properly support multiple languages. The infrastructure piece is there for that. I think app devs can decide how they want to support multiple languages in the meantime, until we decide on how we want to support multiple languages.

Yeah, that makes sense!

I'm not a big fan of the `FormattedMessage` component API. What do you think about creating a component called `Text` that'd be used like this:

```JSX
<Text id="boilerplate.components.Component.id">
  Some Text Here
</Text>
```

Which gets rendered into 

```JSX
<FormattedMessage
  id="boilerplate.components.Component.id"
  defaultMessage={`
    Some Text Here
  `}
/>
```

If no ID is specified, I guess it'd simply render

```
<p>Some Text Here</p>
```

mxstbr commented:
Also, we should add this feature to the `README` and add some documentation about it, since it's really important! @oliverturner any input for the text we could put in the README "Features" section?

oliverturner commented:
Excellent stuff! I'll take a look shortly.

benjaminlgt-littlelives commented:
Added locale loading, polyfill for Intl for Safari, user config file, support for polling for webpack, extraction of translation messages and message loading. -- https://github.com/SilentCicero/react-boilerplate/pull/1

@mxstbr Hmm... if we use `<Text />` it's non react-intl standard and so we cant use the official extraction plugin, will have to roll our own.

Truedrog commented:
@mxstbr with `FormattedMessage` you can provide a default message. How is to do it with `Text`?

SilentCicero commented:
@mxstbr so the one issue I have with creating a `Text` component is that `FormattedMessage` actually does a lot more that it seems. The template string used in `defaultMessage` actually allows values to be injected into the string. So when I create my language files, I can have certain values injected as variables. 

Here is an example from the `Footer` component:

```js
          <FormattedMessage
            id="boilerplate.components.Footer.author.message"
            defaultMessage={`
              Made with love by {author}.
            `}
            values={{
              author: <A href="https://twitter.com/mxstbr">Max Stoiber</A>,
            }}
          />
```

Here you can see values are being injected into the template string. This is a fairly critical feature for i18n and I'm not sure how we would support this with the `children` property in a react component.

The `FormattedMessage` component provides a lot of facility: 
https://github.com/yahoo/react-intl/wiki/Components#formattedmessage

I understand it seems slightly bulky, but I've worked in four frameworks on i18n, this is lean and mean considering React's design. 

We should definitely make additions to the docs as well.

justingreenberg commented:
great work guys :) this is awesome! one thing i would suggest is using `defineMessages` for message definitions. the problem with providing definitions via component (ie FormattedMessage) is that you will not have access to those messages for imperative formatting... for example:

```js
import Helmet from 'react-helmet'
import { FormattedMessage } from 'react-intl'

export default const AboutPage = () => 
  <div>
    <Helmet title="About" /> // rogue string!
    <h1>
      <FormattedMessage
        id="page.about.title"
        defaultMessage="About" />
    </h1>
  </div>
```

in the example above, how would you pass the formatted title string to Helmet component?

a more flexible approach would be:

```js
import Helmet from 'react-helmet'
import {
  defineMessages,
  FormattedMessage,
  injectIntl,
  intlShape
} from 'react-intl'


const messages = defineMessages({
  title: {
    id: 'page.about.title',
    description: 'Title for About Page',
    defaultMessage: 'About'
  }
})

const AboutPage = ({ intl: { formatMessage } }) => 
  <div>
    <Helmet title={formatMessage(messages.title)} />
    <h1>
      <FormattedMessage {...messages.title} />
    </h1>
  </div>

AboutPage.propTypes = {
  intl: intlShape
}

export default injectIntl(AboutPage)
```

this is obviously a contrived example... a more common use case would be formatting title attributes for ARIA, passing a formatted currency string to third party component, etc

SilentCicero commented:
@justingreenberg interesting approach, I think in general your right, have you checked out: https://github.com/SilentCicero/react-boilerplate/pull/1 

Can you make a PR to show us what your thinking, and also considering the @benjaminlgt-littlelives pull in your design.

I'm into this, however, it adds another layer of complexity that I'm not sure is needed in every case. Perhaps it is the right approach though.

So we define all the messages/ids/defaults at the top of each component, then pass those through in the Component properties using them in simple formatted messages within the component itself. Cool!

I'm into this. Your convincing me. Perhaps you can make a PR to benjamins PR once Max takes a look at Benjamins PR lolol. Chain of custody here.

justingreenberg commented:
https://github.com/SilentCicero/react-boilerplate/pull/1LGTM 👍 

one comment... i wouldn't polyfill `intl` [using a script tag](https://github.com/SilentCicero/react-boilerplate/pull/1/files#diff-b9707fe88797285e93483456d1389b61R18), instead i would recommend adding [the intl polyfill](https://www.npmjs.com/package/intl) as dependency and using webpack to create chunk and load at runtime if necessary:

we just have to [check for polyfill before rendering](https://github.com/mxstbr/react-boilerplate/blob/master/app/app.js#L61)...
```js
const render = () => ReactDOM.render(/* ... */) // make render lazy

if (!global.Intl) {
  System.import('intl').then(render)
} else {
  render()
}
```

i'm swamped today but am happy to answer any questions but prob won't be able to create PR until later tonight or tomorrow, just wanted to provide some feedback based on my experience with react-intl :)

justingreenberg commented:
> So we define all the messages/ids/defaults at the top of each component, then pass those through in the Component properties using them in simple formatted messages within the component itself. Cool!

@SilentCicero yes, and it's super flexible! sometimes it even makes sense to create separate file... for example, using the same string in many components throughout your application, instead of having to remember the id, you can define common messages ie `app.messages.js` in root and import where needed (as opposed to having to remember the id of a specific translation)

```js
// core.messages.js
import { defineMessages } from 'react-intl'

export const core = defineMessages({
  home: {
    id: 'core.home.label',
    description: 'Reusable label for home',
    defaultMessage: 'Home'
  }
})
```

```js
// components/links/HomeLink.js
import { core } from '../../core.messages.js'
import { Link } from 'react-router'
import { FormattedMessage } from 'react-intl'

export default const HomeLink = () => 
  <Link to='/'>
    <FormattedMessage {...core.home} />
  </Link>
```

```js
// components/pages/HomePage.js
import { core } from '../../core.messages.js'
import { FormattedMessage } from 'react-intl'

export default const HomePage = ({ children }) =>
  <div>
    <h1>
      <FormattedMessage {...core.home} />
    </h1>
    {children}
  </div>
```

mxstbr commented:
Wrote a comment over at @benjaminlgt-littlelives' PR, will have to think about @justingreenberg's approach for a bit.

What about having a file per container/component called `text.js` (can't think of a better name atm) that contains this information:

```JS
// containers/HomePage/text.js
import { defineMessages } from 'react-intl'

export default defineMessages({
  home: {
    id: 'core.home.label',
    description: 'Reusable label for home',
    defaultMessage: 'Home'
  }
})
```

And then in the Homepage you could go:

```JSX
// components/pages/HomePage.js
import text from './text';
import { FormattedMessage } from 'react-intl';

export default const HomePage = ({ children }) =>
  <div>
    <h1>
      <FormattedMessage {...text} />
    </h1>
    {children}
  </div>
```

justingreenberg commented:
@mxstbr lol great minds think alike :)

mxstbr commented:
No, I saw your comment and took inspiration from it, love that approach, but instead of having a global file I'd much rather have a file per container!

SilentCicero commented:
@mxstbr I think `messages.js` would be most appropriate instead of `text.js`

I'm into breaking messages into their own file for export.

Then the module will import the `messages.js` file and use it in the component.

Something to think about when we are scaffolding. But I like breaking the language up into a separate file, just like selectors/sagas etc.

This is going to be great !

mxstbr commented:
Yes exactly, awesome @SilentCicero!

SilentCicero commented:
@justingreenberg how do we want to proceed with the above changes? Can you make the changes @mxstbr wants to your pull on my PR, then I'll switch everything over to separated `messages.js` files.

SilentCicero commented:
@mxstbr all tests pass, `traviscs/appveyor` white space error is also fixed, `cheap-source-mapping` in `webpack.dev`, all i18n messages are isolated in `messages.js` files in each component, intl provider now supports multiple languages. Scaffolding provides an i18n option for `container` and `component`.

Once this is fixed, you can merge the `react-optimize` https://github.com/mxstbr/react-boilerplate/pull/364 and even font loader refactor https://github.com/mxstbr/react-boilerplate/pull/409 PR's @mxstbr, as they should now pass the `traviscr/appvayor` tests if they merge this PR.

mxstbr commented:
This looks very solid, thanks so much!

I need to test it before I can merge it—I won't have the time to review it properly until after this weekend, I'm sorry for the delay here, I'll get it done asap!

SilentCicero commented:
@mxstbr no worries, I think in general it's the right approach, we may want to readdress transitioning between languages later on.

azampagl commented:
@SilentCicero thanks for the updates. I'll re-test #364 when @mxstbr is done with React Europe and merges this branch.

benjaminlgt-littlelives commented:
@SilentCicero I dont use a .babelrc file, what error do you get running it?

SilentCicero commented:
@benjaminlgt-littlelives hmm, interesting. I get a strange error. 

```
/home/nick/github/react-boilerplate-repo/app/i18n.js:2
import { addLocaleData } from 'react-intl';
^^^^^^

SyntaxError: Unexpected token import
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:387:25)
    at loader (/home/nick/github/react-boilerplate-repo/node_modules/babel-register/lib/node.js:158:5)
    at Object.require.extensions.(anonymous function) [as .js] (/home/nick/github/react-boilerplate-repo/node_modules/babel-register/lib/node.js:168:7)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (extractIntl.js:7:14)
    at Module._compile (module.js:413:34)

npm ERR! Linux 4.0.0-040000-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "extractIntl"
npm ERR! node v5.6.0
npm ERR! npm  v3.8.5
npm ERR! code ELIFECYCLE
npm ERR! react-boilerplate@3.0.0 extractIntl: `babel-node ./internals/scripts/extractIntl.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the react-boilerplate@3.0.0 extractIntl script 'babel-node ./internals/scripts/extractIntl.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the react-boilerplate package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     babel-node ./internals/scripts/extractIntl.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs react-boilerplate
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls react-boilerplate
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/nick/github/react-boilerplate-repo/npm-debug.log
```

Seems like babel is not working properly, which is why I figured .babelrc. Thoughts?

Kielan commented:
@SilentCicero 
I also have a very similar babel error for a different module which I've checked across browsers: 'unexpected token import' as well. maybe rollback on react-optimize? 

Curious to know what's happened.

SilentCicero commented:
@Kielan I haven't integrated `react-optimize`. It seems babel is not working well with `extractIntl`, as token "import" a babel es2015 token is not available.

justingreenberg commented:
@SilentCicero https://github.com/SilentCicero/react-boilerplate/pull/4 https://github.com/SilentCicero/react-boilerplate/pull/5 should fix these issues :)

benjaminlgt-littlelives commented:
@justingreenberg Thanks for hopping in, was tied down past 2 days with work.

SilentCicero commented:
@justingreenberg thanks ! just merged.

mxstbr commented on .emacs:
```diff
@@ -0,0 +1 @@
+(add-hook 'before-save-hook 'delete-trailing-whitespace)
```
Not a big fan of root level configuration files, would love to keep that clean. Can we either move this inside `package.json` or remove it? I don't think many people use emacs with this boilerplate… :grin:

mxstbr commented:
Okay, I'm back now, sorry for the long delay!

Left a small nitpick comment from a cursory glance, also I still don't understand what `internals/scripts/extractIntl.js` is used for and that needs a ton of comments explaining what's going on. (sorry if this was explained somewhere before and I missed it, feel free to copy and paste)

Steps to get this merged:

- [ ] Clean up conflicts, add comments everywhere
- [ ] Thorough review, test the hell out of every part, make sure the solution is solid
- [ ] Documentation about this feature, esp. how to remove it again
- [ ] Merge!

benjaminlgt-littlelives commented:
@mxstbr I commented extractIntl.js before but it somehow got lost in the PRs. I've commented in this gist so it doesnt get lost again :)

https://gist.github.com/benjaminlgt-littlelives/b58949315cbc290228b942cb6e8b00ca

justingreenberg commented:
@mxstbr @SilentCicero @benjaminlgt-littlelives this is a pretty big (and awesome) PR... we should really make an effort not to let it get stale... since this touches so many component and master is a moving target, keeping it in sync will be tough, how can we all pull together and get this merged ASAP?

@mxstbr hope you found your bag dude 🙏

mxstbr commented:
Absolutely, just need some documentation and some more comments in `internals/scripts/extractIntl.js`!

benjaminlgt-littlelives commented:
SilentCicero/react-boilerplate#7

mxstbr commented:
Brilliant! :+1:

SilentCicero commented:
@mxstbr added docs, tested and reviewed. I think we are ready to merge this puppy =D

Can we all take a quick peak at the docs I wrote, see if we need to change any wording/phrasing.

mxstbr commented:
This looks _really_ solid, @oliverturner any last comments about the docs?

SilentCicero commented:
@mxstbr pleasure working with you =D This boiler is getting more rad by the second. I would love to talk sometime, I have big plans for this boiler, but it has to do with some really crazy concepts.. what is the best way to get in touch?

mxstbr commented:
> what is the best way to get in touch?

Twitter DM's probably :grin:

SilentCicero commented:
@mxstbr I'm at @IAmNickDodson :]

breandr commented:
Why does this include jscs?

SilentCicero commented:
@breandr wow, good catch ! I think it was left over from an old merge, just updated the package JSON. We should certainly look it over again. @mxstbr lets compare before we merge, luckily it was just additional dead weight.

mxstbr commented:
I'm taking a look today, just need to get out of this hotel first!

mxstbr commented:
Alright, finally had time to properly review this, again sorry for the delay!

A few nitpicks:

**Code**

- Comment `i18n.js`, explaining what it's for and what each block does. (see `app.js` for inspiration)

**Docs**

- Let's remove the `react-intl` mention from the README. We don't highlight any specific technologies there anymore, and I don't think this should be an exception.

- Rename the docs file to `i18n.md` and change the heading to `Internationalization`. (or something like that) Again, no need to highlight the specific package we use in there. (this also means changing all references to that file)

- Remove "International apps that scale need to support multiple languages." sentence, it doesn't add anything to the doc.

- I don't really like "boiler" as a reference, it sounds weird. Can we change that to "boilerplate" again?

I don't fully understand why we need the `messages.js` files if we have the main translation files? When/where do these get overriden? When do the main translation files get included?

Also, how do I add a new language? "adding an empty generation slot in the `internals/scripts/extractIntl` script" isn't a very good description, I have no idea where I have to add anything there? Isn't it enough to add a new locale to the `i18n.js` `appLocales` variable? If not, can we make it that way?

SilentCicero commented:
@mxstbr these are all great points. Okay, I'll make changes over the next 2-3 days.

So the `messages.js` file contains all the message date for the formatted message components. It lays out i18n objects for the FormattedMessage components. It makes sense to have this, because it contains in one single file, all i18n text structure. This structure is what we use to generate our i18n language JSON files. We need to layout the structure in the component, so that we can design the components with a specific language structure, this way we build components with both default english text, but also with the ability to support any provided translations. So the `messages.js` file seems like a great design pattern as it separates all text/i18n/messages and structure in a single file away from the component. I think it's a healthy and mature separation of concerns, it also makes it easy to remove and work with, especially if we have a fairly large component with a lot of text. 

I would avoid abstracting the FormattedMessage component for now (into a Text like object), (1) because it is easier to understand in the current notation given the `react-intl` docs, (2) there are many properties to the FormattedMessage object that would have to be transfered over, for no reason other than to connect the existing message structures with the component, (3) it's more direct and clear to use the component as it is a "formatted message" and not just "text". It's formatted because there are plug able variables and multi-language support which is formatting the message text.

If you still feel all FormattedMessage components should be abstracted, I would recommend we do it in stages, first working on and accepting this commit, and then adopting the `Text` object it we feel it necessary in a second PR. We should certainly poll the major contributors to see their thoughts, but my opinion is, using the FormattedMessage component/notation is both more direct and explicit than using an abstracted component.

I will work to better explain and comment on adding other languages to the app. I can do a better explanation.

No worries for the delay ;)

SilentCicero commented:
@mxstbr everything should be fixed in the docs. I got lazy (didn't want to explain adding a language), so I just made a plop generator option "language" which will add make all necessary modifications to the `i18n.js` file in addition to adding a shell `[a-z]+.json` language file (so when `extractIntl` is run, it generates an empty lang file in that new language).

To recap, I addressed all doc mods, added a comment header to `i18n.js`, added a plop generator option called "language" to quickly add a language to the app.

So try:
```
npm run generate
``` -> "language" =D

mxstbr commented on docs/js/i18n.md:
````diff
@@ -84,7 +84,17 @@ You can extract all i18n language within each component by running the following
+You can add a language by running the generate command:
+
+```
+  npm run generate
````
Let's change this to `npm run generate language` (plop supports that :wink: ) and remove the "The select…" sentence part below.

mxstbr commented:
> I would avoid abstracting the FormattedMessage component (into a Text like object) for now

I agree with your arguments. :+1: 

If I change some text in a `translations/en.json` file but not in the corresponding `messages.js` file, will it be the changed text in the UI?

SilentCicero commented:
@mxstbr alright, updated the docs CMD line statement.

To answer your question, yes, when you alter the text in the JSON files it will change that text in the app itself. By default the language is set to English. So if you modify the `en.json` file, it will change in the app. All JSON text is loaded in the `i18n.js` file at app start.

benjaminlgt-littlelives commented:
Agreed that we should not use `<text />`, `<FormattedMessage  />` is much better suited and text will just end up being a direct copy of FormattedMessage.

mxstbr commented:
Alright, I know you're going to hate me but I think we should integrate switching of locales into this before we merge it.

The thing with a boilerplate is that the `master` branch is what people use _in production_. If we merge this as–is, there's a really important part missing that everybody using i18n will want to have. (how do we switch out languages from the UI??)

What do you think?

SilentCicero commented:
@justingreenberg @benjaminlgt-littlelives what do we think is the best approach to integrate language toggling at the UI level.

I'm thinking a single, simple state object `Language` with a single property `locale`. This will be handled in the `app/reducers` and `app/actions`. I do not believe it will need any sagas but I could be wrong.

@mxstbr I'll need to experiment with a few approaches. In general, a simple Language object with a defined property `locale` would be the right approach here IMO. We could have a drop-down in the UI, perhaps in the footer where language could be switched. This will require a `LanguageToggle` or `LocaleToggle` component as a simple. Then at a container at which the `LocaleToggle` component would be used, and language could be switched live within the UI.

So I would say, we are looking at at least 5-10 file additions here. One container and one component. And a direct integration of redux state in the App.js, state.Language.locale that will be fed into the `IntlProvider` component.

Thoughts?

breandr commented:
I have no experience with react-intl; could you explain more on `Footer/messages.json` vs `translations/en.json`. Currently, "This project is licensed under the MIT license." is repeated 4 times (perhaps one of those is just placeholder in de.js?). I prefer the approach of keeping the messages with the component and then using export to extract just the english JSON files to hand off to translation team, and then import once translation is completed. Maybe this is exactly what is happening, and `translations/en.json` represents a built/compiled language file? Ideally, no messages should be repeated.

On the subject of loading these messages: the user should not have to download all messages for all languages (which seems to be the case with app/i18n.js) and they should only load the languages necessary for the page/component they are viewing; ie. messages should be lazy-loaded for the language and the page/component/route. Ideally, default messages would always be loaded/bundled, and if the user has a language selected other than default (eg. en), it should asynchronously load the other language file (eg. de) and fallback to the default message for any messages that do not have a translation in de.

mxstbr commented:
*(Also, just noticed this exists—maybe we should use it instead of our custom `extractIntl`? https://github.com/yahoo/babel-plugin-react-intl)*

SilentCicero commented:
@mxstbr agreed, we should investigate that package. For now the extract script should do.

The `react-intl-redux` package should be vetted before we move forward with it. It actually gives us a whole new redux product abstraction (`Provider` component), which I don't feel very comfortable with. Perhaps it's fine, as it it a fairly simple extraction. But naturally I'm skeptical of that design pattern.

If we feel `react-intl-redux` to use unhealthy design patterns, then we should opt for servicing our own redux language and locale design.

@mxstbr @benjaminlgt-littlelives @justingreenberg can you look over the  `react-intl-redux` package. Would love to get thoughts before I move forward on implementation.

mxstbr commented:
> If we feel react-intl-redux to use unhealthy design patterns, then we should opt for servicing our own redux language and locale design.

Sounds reasonable, I'm not a fan of combining the `Provider` and the `IntlProvider` either!

SilentCicero commented:
@mxstbr I think our unix philosophy senses are tingling.. =D

benjaminlgt-littlelives commented:
@SilentCicero I like the implementation you speak of(locale in the store which gets passed to `IntlProvider`), and agree that `Provider` and `IntlProvider` should not be combined, hence I'm in favour of rolling our own. I could help with that but I'm tied up this week.

@breandr Chunked language files are possibly something we can look at in the future, but I don't think it needs to be in the initial release.

en.json is the file you hand off to the translation team to fill in, I do agree that there is some repetition(The english message) we can do away with for the built files we serve to the user. That is something we can look at together with language chunking.

I do however like having messages.js as a separate file since it makes things cleaner.

@mxstbr As mentioned in the initial  issue(https://github.com/mxstbr/react-boilerplate/issues/377#issuecomment-221215529), https://github.com/yahoo/babel-plugin-react-intl is exactly what extractIntl uses behind the scenes. But I added additional functionalities such as per locale language files, sorting so that git diffs are cleaner, and merging with existing language files so that what we don't lose the translations we already have.

mxstbr commented:
> @mxstbr As mentioned in the initial issue(#377 (comment)), https://github.com/yahoo/babel-plugin-react-intl is exactly what extractIntl uses behind the scenes. But I added additional functionalities such as per locale language files, sorting so that git diffs are cleaner, and merging with existing language files so that what we don't lose the translations we already have, and also additional error handling messages as well as verbose logging to the console.

Ah sorry, I missed that, makes sense! Thanks!!

SilentCicero commented:
@benjaminlgt-littlelives @mxstbr working on something now. Should have it up today.

SilentCicero commented:
@mxstbr @benjaminlgt-littlelives @justingreenberg just a preliminary heads up.

I'm gonna post up what I have in a commit in like two hours. We need to talk names/patterns etc. After we discuss this next commit, I'll patch and test what I have. Then I think we are done.

Just an FYI.

breandr commented:
I agree chunked messages should be a future enhancement, but I still don't see the need for a message to be repeated more than once. Will I need to add/edit messages in `app/translations/en.json`,  `app/translations/de.json`, AND `app/components/Footer/messages.js`?

justingreenberg commented:
@breandr `react-intl`'s `defineMessages` singleton acts as the "entry point" for messages; the extractIntl.js script (or babel plugin) builds the translation files for you... devs should never have to directly edit any json translations

breandr commented:
Thanks @justingreenberg, I was hoping this was the answer.

benjaminlgt-littlelives commented:
@SilentCicero Awesome!

@breandr Yes, you do not have to edit the messages more than once, it's existence in the generated files is just to guide your translators, every time extractIntl.js is run it ignores the existing messages in the translation files and pulls the ones from `defineMessages`

SilentCicero commented:
@mxstbr @justingreenberg @breandr @benjaminlgt-littlelives alright, this is a first pass approach here.

So I created a `IntlProvider` abstraction: `LanguageProvider`. This takes in the translated messages from the `i18n.js` file. I included this provider in `app.js` and `templates...app.js`. It has a simple action/reducer setup. This is the default redux state:

```
{
   language: {
      locale: "en",
   },
}
```

The action is fairly simple: `CHANGE_LOCALE`. This takes in a single two character locale and changes it. No `redux-sagas` are need for this.

`LanguageProvider` is the main component here that connects the redux state to react-intl. It's a single component that meets that single concern. Note, I also added the `LanguageProvider` and `i18n.js` and `translactions/en.json` to the `clean.js` script.

Then I added some basic unnopinionated `Toggle` and `ToggleOption` components. These are for the `LocaleToggle` which is a container. The Toggle component takes in prop `values` and `messages`, and then makes a single `select` element out of that.

Then I added a `LocaleToggle` container, which uses the Toggle component, and the actions/reducers from `LanguageProvider` to create a simple locale toggle feature, so you can switch between languages within the UI. The LocaleToggle preps and assembles the Toggle with a set of i18n messages and languages.

Would like to get thoughts on this.

Anyway, the end result is....... we now can switch languages in the UI, via redux state !!!

So lets get out opinions together on names and approaches, and commit this.

justingreenberg commented:
@SilentCicero awesome work dude 👍 will throughly review tonight!

justingreenberg commented on app/components/ToggleOption/index.js:
```diff
@@ -0,0 +1,23 @@
+function ToggleOption(props) {
+  return (
+    <option value={props.value}>
+      <FormattedMessage {...props.message} />
```
this is throwing an error in browser...
```
warning.js:44 Warning: Only strings and numbers are supported as <option> children.
```
we will need to use `injectIntl` and formatMessage api here to pass string

SilentCicero replied:
Great find, okay I'll add this. I was wondering what the approach should be here.

SilentCicero replied:
@justingreenberg can you make a PR to fix this? I wont have time this weekend.

justingreenberg replied:
of course! normally i would have opened PRs for these but i was beyond exhausted last night, just wanted to review everything as promised :) will try to find some time later today to address these

justingreenberg commented on app/containers/LanguageProvider/reducer.js:
```diff
@@ -0,0 +1,26 @@
+} from './constants';
+
+const initialState = fromJS({
+  locale: 'en',
```
maybe we should try to grab browser locale? ie

```js
window.navigator.userLanguage || window.navigator.language;
```

if we do try to detect user locale, we will need to check it against our supported languages (`appLocales` export from `i18n`)

SilentCicero replied:
@justingreenberg if you can make a PR for this as well, that would be a great help.

justingreenberg commented on app/i18n.js:
```diff
@@ -0,0 +1,35 @@
+
+import enLocaleData from 'react-intl/locale-data/en';
+import deLocaleData from 'react-intl/locale-data/de';
+
```
maybe we should export a `defaultLocale` here and replace hardcoded 'en' strings in provider setup

SilentCicero replied:
I would say we could approach this later. For now being clearly expressive about the locales seems fine to me.

mxstbr commented:
> maybe we should try to grab browser locale?

Let's get that in, then I'll review it one final time and we finally merge and release this?! :shipit:

justingreenberg commented:
hey max, i looked into browser locale detection and i don't think it's worth the cost... detecting locale using browser api is problematic due to vendor inconsistencies and would require a function similar to https://github.com/maxogden/browser-locale/blob/master/index.js

the language code returned from browser would then have to parsed and normalized for our own i18n implementation, ie. "en-US" and "en-UK" ⟶ "en"... finally, we would have to check final language against our supported locales... 

there is an alternative technique that involves sniffing request headers, which we can implement pretty easily once ssr is baked in.. let me know what you think!

this is pretty much good to go, just needs some polish and could use some ui direction.. any thoughts on where you want to place language toggle? maybe in footer where the MIT license is?

 - [ ] clean up toggle ui
 - [ ] complete `de` translations (def need your help with this one 😄)

mxstbr commented:
> there is an alternative technique that involves sniffing request headers, which we can implement pretty easily once ssr is baked in..

Alright, let's go with that then and leave it english-default with a switch for now!

> maybe in footer where the MIT license is?

That sounds reasonable!

> complete de translations

Will look into it :grin:

SilentCicero commented:
@mxstbr I put the LocaleToggle in the footer beside the MIT license. All we need is the German translation.

mxstbr commented:
On it, will do tomorrow evening! :+1:

rog commented:
what's the status of this PR?

SilentCicero commented:
@rog she is close, @mxstbr needs to do german translations, and I need to do push more test coverage of the LocaleToggle component. Then we good to go.

rog commented:
can I help with something? @mxstbr @SilentCicero

mxstbr commented:
@rog do you speak another language apart from English and German? More translations can't really hurt :grin:

mxstbr commented:
I don't want to be a downer, but the text on the Features page is a bit messed up:

![screen shot 2016-06-23 at 12 02 57](https://cloud.githubusercontent.com/assets/7525670/16299425/7db0d5c8-393a-11e6-81cd-ea5309a6273c.png)

mxstbr commented:
[German translations PR](https://github.com/SilentCicero/react-boilerplate/pull/9) is done! One more thing, any chance we could use `module.hot` (or something like that) to hot-reload the translation JSON files?

rog commented:
Yeah, I will make spanish translations.

SilentCicero commented:
@mxstbr yeah, lets jazz the feature page text up. Any suggestions? If not, I'll just make it close to the others. @rog awesome.

Re: module.hot, I'm not totally sure on how to make the JSON module hot. Perhaps there is a way. Anyone have any ideas?

mxstbr commented:
> @mxstbr yeah, lets jazz the feature page text up. Any suggestions?

Just copy and paste the README HTML code, that's what we originally did but somehow some sections are now duplicated and some are missing.

Honestly no idea, but non-hot-reloadable text is kind-of inoptimal, right? :confused: maybe we could do something like we do for the reducers in `app.js`?

SilentCicero commented:
@mxstbr okay, fixed most of that. Where is the app.js hot module replacement code? I cant seem to find it.

mxstbr commented:
Ah damn, that's in `store.js`, I'm sorry!

SilentCicero commented:
@mxstbr I'll patch and test, should be up in 3 hours :-)

SilentCicero commented:
@mxstbr I gave it ago, you still need to hit refresh on the browser. I'm close. Any help would be welcome @justingreenberg @rog @benjaminlgt-littlelives -- check the latest commit for HMR for the language i18n JSON files.

benjaminlgt-littlelives commented:
@SilentCicero Made HMR work, https://github.com/SilentCicero/react-boilerplate/pull/10

SilentCicero commented:
@benjaminlgt-littlelives awesome ! Is it happening @mxstbr is this it? :-]

mxstbr commented:
Doing one more review shortly to make 100% sure this is exactly what we want, including cleaning and setting up a new project!

mxstbr commented:
The text on the features page still repeats itself:

![screen shot 2016-06-27 at 14 05 25](https://cloud.githubusercontent.com/assets/7525670/16379172/41e6c56e-3c70-11e6-8a1c-01b4746f4887.png)

SilentCicero commented:
@mxstbr will fix, any other issues?

SilentCicero commented:
@mxstbr fixed text, all tests pass. Anything else?

mxstbr commented:
When trying to clean, this is what I get:

```Sh
➜  rb-test git:(i18n) npm run clean

> react-boilerplate@3.0.0 clean /Users/a2517/sites/experiments/rb-test
> shjs ./internals/scripts/clean.js

Cleanup started...cp: no such file or directory: internals/templates/notFoundPage/notFoundPageMessages.js
cp: no such file or directory: internals/templates/notFoundPage.js
cp: no such file or directory: internals/templates/homePage.js
mkdir: path already exists: app/translations
```

SilentCicero commented:
@mxstbr will fix up the clean script. Should be done today or early tomorrow.

benjaminlgt-littlelives commented:
@SilentCicero Done - https://github.com/SilentCicero/react-boilerplate/pull/11

SilentCicero commented:
@mxstbr should be good to go here.

benjaminlgt commented:
@mxstbr Does it look good to you?

mxstbr commented:
Sorry for the delay here, was on holidays! Reviewing at the moment, then we'll merge this!

mxstbr commented:
**Blocking** (has to be done before we merge)

- [ ] In German, the "Features" page has one title and description inverted (probably my mistake):

  ![screen shot 2016-07-13 at 09 46 49](https://cloud.githubusercontent.com/assets/7525670/16795637/ca297eec-48de-11e6-8fed-6154847abf02.png)

- [ ] Insert empty last line into the JSON files when running `npm run extractIntl`, otherwise there's unnecessary git changes.

- [ ] Should we run `npm run extractIntl` after generating a language so the `.json` is filled with something?

- [ ] `npm run extractIntl` should show progress indicator like other scripts

- [ ] After running `npm run clean` the `i18n.js` still references german a few times, but not everywhere:

  ![screen shot 2016-07-13 at 10 06 43](https://cloud.githubusercontent.com/assets/7525670/16796186/93e8c36c-48e1-11e6-9591-20e3d7537064.png)

- [ ] When cleaning completely replace the `LanguageProvider`, like the other containers. Simply leaving it be won't work since people might've played around/changed the example, and if they changed/broke the `LanguageProvider` they'll still have those changes after cleaning, which is a no-go!

----

**Todo** (can be done after the merge, would be nice if we got it done before though)

- [ ] When generating a language we insert an empty line after the injected locale (we have that bug somewhere else too I think)

  ![screen shot 2016-07-13 at 09 54 28](https://cloud.githubusercontent.com/assets/7525670/16795876/f575307c-48df-11e6-8536-745305908f4a.png)

- [ ] Test `languageProviderReducer` switches locale when a type of `CHANGE_LOCALE` comes in
- [ ] Test `selectLocale` selector

----

Unrelated, I _love_ the `npm run extractIntl` output 😍 

<img width="651" alt="screen shot 2016-07-13 at 10 01 05" src="https://cloud.githubusercontent.com/assets/7525670/16796037/c5da5d0a-48e0-11e6-8d7c-0bf3a45c1f05.png">

Also, the HMR for the text is amazing, good job on that!

**I'm not going to merge any big changes into `master` until we merge this in an effort to finally land this!** (/cc @gihrig) Sorry for taking so long, you should not have to do any more rebases!

joeblew99 commented:
:) +1 Almost here. Really awesome work on this

benjaminlgt-littlelives commented:
1 and 2 are done - https://github.com/SilentCicero/react-boilerplate/pull/12

@mxstbr Sorry what do you mean by a progress indicator?

I'm not too familiar with running npm functions from within plop maybe somebody can help with that.

mxstbr commented:
> I'm not too familiar with running npm functions from within plop maybe somebody can help with that.

Take a look at the setup.js script: https://github.com/mxstbr/react-boilerplate/blob/master/internals/scripts/setup.js#L35-L38

Used something like this:

```JS
var animateProgress = require('./helpers/progress');

var interval = animateProgress('Initialising new repository');
process.stdout.write('Initialising new repository');
callback = () => {
  clearInterval(interval);
}
```

bang88 commented:
Can't waiting this feature landing.

SilentCicero commented:
@benjaminlgt-littlelives @justingreenberg any help on this to get it too the finish line would be great ! Dont have time right now due to work. Just make PR's on my repo if you can.

justingreenberg commented:
Ok I will try to wrap this up when I get home, we just need to sync with master and add loading indicator if I'm not mistaken?

> On Jul 20, 2016, at 6:43 PM, Nick Dodson <notifications@github.com> wrote:
> 
> @benjaminlgt-littlelives @justingreenberg any help on this to get it too the finish line would be great ! Dont have time right now due to work. Just make PR's on my repo if you can.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

SilentCicero commented:
@justingreenberg awesome, yes I believe that is all, thanks again

justingreenberg commented:
@SilentCicero no problem, working on this now... rebasing looks a little gnarly so i am going to step through base changes one by one, will submit pr in ~ an hour or two

justingreenberg commented:
@SilentCicero please see https://github.com/SilentCicero/react-boilerplate/pull/13

@mxstbr the animation is so fast you can't really see lol... but is this what you had in mind?

![extract-intl](https://cloud.githubusercontent.com/assets/1539088/17039927/e4432368-4f6a-11e6-9ea4-70ab97e73783.gif)

SilentCicero commented:
@justingreenberg I love how you made a video. Can I hire you, can you come to canada?

mxstbr commented:
@justingreenberg yeah, kind of haha!

Have we fixed `npm run clean` yet? As far as I can tell that's the last thing missing to merge this.

chaintan17 commented:
this is... wait for it... legendary pull request, ever :-D
Anyway, i would like to contribute to this pull request, however, it's now conflict to master and i cannot retrieve it locally... Can someone who have write access update it to match master branch please?

chaintan17 commented:
I'm working with integration to translation management tools (PhraseApp)
And it expect the translation file to be this format 
```json
{
  "app_title":"My Software Project",
  "greeting":"Hello 'User'!\nHow are you doing today?"
}
```
ref. https://phraseapp.com/docs/guides/formats/react-intl-simple-json/

Like @breandr mentioned earlier, do we really need to export `defaultMessage` to every language?
I think it's more nice to keep translation json to only have a `key` and translated message.
And let `i18n.js` do the job, transforming translation json file to expected react-intl format.

What do you guys think? cc @justingreenberg

surzhik commented:
Hello.
Pls, give me the way.
How to load and use message locales inside containers?
Without duplication it in json in translations folder.
My app will be big with about 30 complex modules.
Putting all messages into one json - total mess.
Any idea?
Thanks.

gihrig commented:
I believe what you are asking is not a simple task and is really beyond the scope of this project.

See [andyearnshaw/Intl.js](https://github.com/andyearnshaw/Intl.js) and [yahoo/react-intl ](https://github.com/yahoo/react-intl)for details.

If you take this path be sure to experiment on a small scale before committing a lot of work to it.

*Once you start down this path, you're on your own. This project will not be able to support you in whatever happens*.

But are you _sure_ you want to exchange one large translation file for ~30 smaller ones? Seems to me from a translator view, one large file is better than ~30 small ones. Also much easier for you as a dev to maintain.

chaintan17 commented:
I create pull request for shrinking translation message js files
and support PhraseApp intl import
Some work need to be done, please review
https://github.com/mxstbr/react-boilerplate/pull/931

surzhik commented:
@gihrig, it is not question about files translation. It is question about management and separate modules development. 
I did own solution. It is not perfect, but works for me.
First of all I declaring that all components must have translation folder with locales files. Files can be empty, but must exist. Then add location property to createRoutes object:
`return [
    {
      path: '/',
      name: 'My Home',
      location: 'containers/HomePage',
`
In app.js making array with target locales:
`
const traverse = (o) => {
  let locations = [];
  for (const i of o) {
    if (!!o[i] && typeof(o[i])==="object") {
      if (o[i].location)
        locations.push(o[i].location);
      traverse(o[i]);
    }
  }
  return locations;
};

// Import i18n messages
//import { translationMessages } from './i18n';
import { translationMessagesExtend } from './i18n';

const translationMessages = translationMessagesExtend(traverse(rootRoute.childRoutes));
`
translationMessagesExtend new function in i18n.js
`
export const translationMessagesExtend  = (locations) => {
  const localeFiles = {};
  locations.push('');
  for (const locale of appLocales) {
      localeFiles[locale] = [];
      for (const location of locations) {
         let path = location ===""?location + "translations/" + locale:location + "/translations/" + locale ;
         let req = require("./" + path + ".json")
         localeFiles[locale].push(...req);
     }
  }
 for (const locale of appLocales) {
    localeFiles[locale] = formatTranslationMessages(localeFiles[locale])
  }
  return  localeFiles
};
`
Also putting current locale messages into store. LanguageProvider/reducer.js
`
import { fromJS } from 'immutable';
import {
  CHANGE_LOCALE,
  LOAD_LOCALES
} from './constants';


import { selectLocaleOnLoad } from './selectors';

const initialState = fromJS({
  locale: selectLocaleOnLoad(),
  messages:{}
});

function languageProviderReducer(state = initialState, action) {
  switch (action.type) {
    case CHANGE_LOCALE:
      return state
        .set('locale', action.locale);
    case LOAD_LOCALES:
      return state
        .set('messages', action.messages);
    default:
      return state;
  }
}
export default languageProviderReducer;
`
It gives me possibility do not load default messages into each module. Here is selectorЫ from LanguageProvider/selectors:
`
import { createSelector } from 'reselect';
import { defineMessages } from 'react-intl';

let localStorage

// If we're testing, use a local storage polyfill
if (global.process && process.env.NODE_ENV === 'test') {
  localStorage = require('localStorage')
} else {
  // If not, use the browser one
  localStorage = global.window.localStorage
}

const selectLanguage = () => state => state.get('language');

const deCompileMsgs = (messages) => {

  const formattedMessages = {};
  for (const message in  messages) {
    let mForCup = message.split(".");
    for(const i in mForCup){
      mForCup[i] = mForCup[i].charAt(0).toUpperCase() + mForCup[i].substr(1);
    }
    formattedMessages[mForCup.join("")] = {id:message, defaultMessage:messages[message]};
  }

  return defineMessages(formattedMessages);
};


const selectLocale = () => createSelector(
  selectLanguage(),
  (languageState) => languageState.get('locale')
);
const selectMessages = () => createSelector(
  selectLanguage(),
  (languageState) => deCompileMsgs(languageState.get('messages'))
);
const selectLocaleOnLoad = () => {
  return localStorage.getItem("locale")?localStorage.getItem("locale"):"en"

}

export {
  selectLanguage,
  selectLocale,
  selectLocaleOnLoad,
  selectMessages
};
`
And in module:
`import { selectMessages, selectLocale } from '../LanguageProvider/selectors';
/*...*/
 return (
      <div className='nav'>
          <FormattedMessage {...this.props.messages.CoreolaContainersNavbarTitle}/></h1>
      </div>
    )
/*...*/
const mapStateToProps = createStructuredSelector({
  data: selectFormState(),
  locale: selectLocale(),
  messages: selectMessages()
});
`
CoreolaContainersNavbarTitle produced from id:
`
{
    "id": "coreola.containers.Navbar.title",
    "defaultMessage": "Coreola"
  }
`
So, you don't need to extend object with name.

Thats it.

apinnecke commented:
@chaintan17 Big thanks for your work! Really looking forward to see this PR merged to get react-boilerplate ready for PhraseApp! 💪  cc @schubidu
